Add bluetooth buttons and bluetooth headphone output toggle

ChisSoc opened the pull request:
The pull requst to the two feature proposals #1192  and #1193 . This ended in a single pull request as both features are closely intertwined and have some cross-feature that are best tested if both are present in the same branch. Makes life a little easier.

It got delayed through more testing and by adding support for easier setup and install by the user. 

The web interface addition is coded with Ajax to accurately reflect current bluetooth status.

s-martin commented on components/bluetooth-sink-switch/README.md:
```diff
@@ -0,0 +1,276 @@
+**Status LED**
+
+An optional status LED will be turned on if the audio sink is set to bluetooth. If a toggle command is issued, but no bluetooth device is connected, the LED will blink three times. Looks very neat, if you have a button with integrated LED. 
+
```
I think it would make sense to use GPIO control (which already has led function) or gpio_settings.ini for this setting. We introduced gpio_setttings.ini to have one central place for GPIO configuration

ChisSoc replied:
Do you mean, 
-  Use the gpio_settings.ini to read the GPIO number for the LED from? Is there an API that already does this for me, which I could use?
- Use the gpio API to switch an LED on/off? 

In any case, when using gpio_settings.ini, an LED device is opened in the process gpio_control.py. But I want to control the LED form bt-sink-switch.py (to avoid having to poll bt-sink-switch status from gpio_control.py). This will potentially lead to a resource conflict of the LED pin. How do I resolve that?

s-martin replied:
@veloxidSchweiz, do you have an idea?

s-martin commented on components/bluetooth-sink-switch/README.md:
```diff
@@ -0,0 +1,276 @@
+
+**Volume attenuation**
+
+Speakers and Headphones can have very different maximum volume levels. This sometimes leads to very strong volume level changes when switching between speakers and headphones. Restructing the maximum volume with the Phoniebox-integrated max-volume setting does no yield the desired effect, as this is a single setting and does not differentiate between different audio sinks.
```
"Restricting"

s-martin commented on components/bluetooth-sink-switch/bt-sink-switch.py:
```diff
@@ -0,0 +1,186 @@
+
+# If called as script, this variable will set GPIO number of the LED to reflect sink status
+# Uses BCM GPIO numbering, i.e. 'led_pin = 6' means GPIO6
+# Set 'led_pin=None' to disable LED support (and no GPIO pin is blocked in this case)
```
See comment in readme

s-martin commented on components/bluetooth-sink-switch/bt-sink-switch.py:
```diff
@@ -0,0 +1,186 @@
+        logger.error("Invalid command. Doing nothing.")
+        return
+
+    # Rudimentary check if LED pin request is valid GPIO pin number
```
See comment in readme

s-martin commented on components/control/bluetooth-buttons/README.md:
````diff
@@ -0,0 +1,82 @@
+```
+- Go into the source code and adjust these lines for desired behaviour
+~~~python
+                if event.code == bt_keycode_play:
````
Would it make sense to have a config file? Editing source files has the disadvantages that inexperienced users may accidentally break the code and that the git workspace now contains modified files.

ChisSoc replied:
Firstly, key codes a standardized and existing key codes should work with any headset. I have mapped all key codes of the Bluetooth headphones I could lay my fingers in the script. If somebody has more, they could add them and make a pull request. The script should work unmodified for the vast majority of use cases.  

Cross-mapping buttons to a different behaviour is of course possible, but not the main use case. 

Inexperienced uses get a chance to learn some Python :-) Having a modified script file is of course not ideal, but should no happen unless you do a cross-mapping. 

A generic config file is a lot of work to set up. I see this as possible future enhancement.

s-martin replied:
Agreed, leaving that as future improvement.

s-martin commented on htdocs/lang/lang-de-DE.php:
```diff
@@ -1,6 +1,5 @@
 <?php
```
Is it intentional that all comments in the file have been removed?

s-martin commented:
Hi, thanks for putting all the work adding this new feature! Very impressive! 
Great that you provided this detailed documentation.

s-martin commented:
For reference, here's a description for mopidy
https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/284#issuecomment-553353467

ChisSoc commented:
Thanks for your good feedback! I'll stay tuned to see what I can do about better LED integration.

MiczFlor commented on scripts/playout_controls.sh:
```diff
@@ -994,6 +995,10 @@ case $COMMAND in
         ;;
+    bluetoothtoggle)
+        if [ "${DEBUG_playout_controls_sh}" == "TRUE" ]; then echo "   ${COMMAND}" >> ${PATHDATA}/../logs/debug.log; fi
+	$PATHDATA/../components/bluetooth-sink-switch/bt-sink-switch.py $VALUE
```
@ChisSoc 
Hi Chris, you really put a lot of thought and work into this!!! Kudos.
Looking at the code in github, it seems that here is an incoherence with the indents. Is that possible? 
Is this code you have running?

ChisSoc replied:
Hi Micz,

thank you very much! I was actually worried, I had put the new feature too deeply into the code base. Glad you like it! 

And also thanks for pointing out my misspelled user name! Actually, it is Chris, and should have been ChrisSoc. A lot of work had been done late at night :-) Solves the mystery why I only could log in with my e-mail.

This has most likely been said before, but cannot be said often enough: This is a very inspiring project! I finished my Box in May 2020 but with Corona and all the implementation of this feature got delayed and the work to bring this feature up to integration quality even more though. I will post pictures soon.

Now, to the issue: The indentation was due to a tabbed indent instead of space characters. I need to apologize for this slip up. I was probably due to me coding some of this directly via SSH on the RPi with Emacs only partially setup to my usual configuration. It is fixed! Bye the way: the code was running on my two RPis - propably because it is a Shell script and not Python? It is working as expected with the fix.

~~~
$ ./playout_controls.sh -c=bluetoothtoggle -v=toggle
No bluetooth device connected. Defaulting to "Output 1".
Switched audio sink to "Output 1"
~~~

MiczFlor commented on scripts/startup-scripts.sh:
```diff
@@ -52,3 +52,15 @@ mpc rescan
 fi
+
+#######################
+# Default audio sink to speakers (instead of bluetooth device) irrespective of setting at shutdown
```
@ChisSoc 
reading "Default audio sink", I thought you meant "sync". But then I found that there are the terms "sinking" and "sourcing" and - not that I understood it all - that might be the reference. Is it?

MiczFlor reviewed (commented):
Hi @ChisSoc,
I asked two questions in the code, calling you "Chris". My apologies. Only now I see the "Chis".
This is a great contribution, thank you. Kudos!!!

MPierer commented:
Hi ChisSoc and MiczFlor.
That's a great addition! I saw the limitation about mopidy on multi-output switching.
I don't want to hijack here, but is there a reason why ALSA is used instead of PulseAudio? PulseAudio could unify mpd and mopidy, and on-the-fly output switching (e.g. card1 headphone jack, card2 speakers, bluetooth) could be achieved even with command line commands (pactl move-sink-input, ...)
Maybe something to consider (or I give it a try myself when time permits)

s-martin commented:
> I don't want to hijack here, but is there a reason why ALSA is used instead of PulseAudio? PulseAudio could unify mpd and mopidy, and on-the-fly output switching (e.g. card1 headphone jack, card2 speakers, bluetooth) could be achieved even with command line commands (pactl move-sink-input, ...)

https://github.com/MiczFlor/RPi-Jukebox-RFID/discussions/1208

In this discussion it was pointed out that Pulseaudio needs X11 as dependency, which is quite heavy on Pi Zero.
If you have ideas, this is more than welcome.

ChisSoc commented:
Hi @MiczFlor,

I replied to you comments directly in the code. Since I completely agree to you comments I fixed it and marked to conversation as "resolved". I am not sure if you get an automated notification it I do that on Github. While being accustomed to coding and using Git in large teams, Github is new to me. :-) 

So here is my reply just in case:
thank you very much! I was actually worried, I had put the new feature too deeply into the code base. Glad you like it!

Now, to the issue: The indentation was due to a tabbed indent instead of space characters. I need to apologize for this slip up. I was probably due to me coding some of this directly via SSH on the RPi with Emacs only partially setup to my usual configuration. It is fixed! Bye the way: the code was running on my two RPis - probably because it is a Shell script and not Python? It is working as expected with the fix.

I also changed "default audio sink" to "default audio output".

greg4b commented:
How can I experience this long awaited feature?

You may excuse my question  - I can't see that in Version 2.2 - 305325d - master

Thanks,

Gregor

s-martin commented:
@MiczFlor, do you have any open issues? Otherwise we could merge it?

MiczFlor commented:
Hi @ChisSoc @arne123 
@s-martin 
This bluethooth feature is awesome. Because I merged @arne123 refactoring of `components/gpio_control/function_calls.py` into classes before, now there is a conflict. Could the two of you rub your brains together and make sparks fly to resolve this conflic?

ChisSoc commented:
Sure. I will look into it. But it will probably be the weekend before I really get a chance to do something about it.

ChisSoc commented:
Merge dand tested. :-)

MiczFlor commented:
@ChisSoc thanks for the quick merger. I will merge it with develop now.
